src/Endpoints/class-base-api-proxy.php

```diff
@@ -0,0 +1,144 @@
+<?php
+/**
+ * Endpoints: Base API proxy endpoint class for all API proxy endpoints
+ *
+ * @package Parsely
+ * @since   3.4.0
+ */
+
+declare(strict_types=1);
+
+namespace Parsely\Endpoints;
+
+use Parsely\Parsely;
+use Parsely\RemoteAPI\Proxy;
+use stdClass;
+use WP_Error;
+use WP_REST_Request;
+use WP_REST_Server;
+
+/**
+ * Configures a REST API endpoint for use.
+ */
+abstract class Base_API_Proxy {
+	/**
+	 * Parsely object instance.
+	 *
+	 * @var Parsely
+	 */
+	private $parsely;
+
+	/**
+	 * Proxy object which does the actual calls to the Parse.ly API.
+	 *
+	 * @var Proxy
+	 */
+	private $proxy;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param Parsely $parsely Instance of Parsely class.
+	 * @param Proxy   $proxy   Proxy object which does the actual calls to the
+	 *                         Parse.ly API.
+	 */
+	public function __construct( Parsely $parsely, Proxy $proxy ) {
+		$this->parsely = $parsely;
+		$this->proxy   = $proxy;
+	}
+
+	/**
+	 * Registers the endpoint's WP REST route.
+	 *
+	 * @param string $endpoint The endpoint (e.g. /analytics/posts).
+	 */
+	protected function register_endpoint( string $endpoint ): void {
+		$filter_key = trim( str_replace( '/', '_', $endpoint ), '_' );
+		if ( ! apply_filters( 'wp_parsely_enable_' . $filter_key . '_api_proxy', true ) ) {
+			return;
+		}
+
+		$get_items_args = array(
+			'query' => array(
+				'default'           => array(),
+				'sanitize_callback' => function ( $query ) {
+					// question: how should we sanitize these?
+					return (array) $query;
+				},
+			),
+		);
+
+		$rest_route_args = array(
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'get_items' ),
+				'permission_callback' => array( $this, 'permission_callback' ),
+				'args'                => $get_items_args,
+			),
+		);
+
+		register_rest_route( 'wp-parsely/v1', $endpoint, $rest_route_args );
+	}
+
+	/**
+	 * Determines if there are enough permissions to call the endpoint.
+	 *
+	 * @return bool
+	 */
+	public function permission_callback(): bool {
+		// Unauthenticated.
+		return true;
```

Comment:
Consider making this an abstract method, so that implementors do not accidentally inherit unauthenticated endpoints.

Reply:
Done.